kernel/arch/dreamcast/hardware/sq.c

```diff
@@ -39,22 +42,55 @@
         QACR1 = (((uintptr_t)(dest1)) >> 24) & 0x1c; \
     } while(0)
 
-static mutex_t sq_mutex = MUTEX_INITIALIZER;
+static mutex_t sq_mutex = RECURSIVE_MUTEX_INITIALIZER;
+
+typedef struct sq_state {
+    void *dest0;
+    void *dest1;
+    mmu_token_t mmu_token;
+} sq_state_t;
 
-static mmu_token_t mmu_token;
+#define SQ_STATE_CACHE_SIZE 8
```

Comment:
```suggestion
#ifndef SQ_STATE_CACHE_SIZE
#define SQ_STATE_CACHE_SIZE 8
#endif
```
This is what I'd do, just so that if some crazy person ever shows up needing 20 layers of SQ locks deep, we can tell them to just pass `-DSQ_STATE_CACHE_SIZE=9999` to `make` when they build KOS... lol.